opencog/scm/opencog/base/utilities.scm

```diff
@@ -287,23 +287,45 @@
 	)
 )
 
+; -----------------------------------------------------------------------
+(define (traverse-roots func)
+"
+  traverse-roots -- Applies func to every root atom in the atomspace.
+
+  The root atoms are those, which have no incoming atoms,
+  located in the atomspace or its ancestors (i.e. visible from the atomspace).
+"
+	(define (is-visible? atom)
+		(member
+			(cog-as atom)
+			(get-atomspace-and-parents
+				(cog-atomspace)
+				`())))
+	(define (get-atomspace-and-parents atomspace res)
+		(if (null? atomspace)
+			res
+			(get-atomspace-and-parents
+				(cog-atomspace-env atomspace)
+				(cons atomspace res))))
```

Comment:
I think the implementation of `get-atomspace-and-parents` can be simplified using `fold` https://srfi.schemers.org/srfi-1/srfi-1.html#FoldUnfoldMap

It's a minor remark, the code looks good anyway.

Reply:
I will check, thanks.

Reply:
It was a good idea. Check #1692 1692